### PR TITLE
Add /close command.

### PR DESCRIPTION
**Pull request by spxtr**
This lets authors and assignees of issues `/close` the issue. Part of the goal of this PR is to be a guide for making new plugins. It's pretty straightforward, and hopefully this is enough for others to start working.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/997)

### CODE REVIEW

File: prow/cmd/hook/main.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ import (
 	"k8s.io/test-infra/prow/kube"
 	"k8s.io/test-infra/prow/plugins"
 
+	_ "k8s.io/test-infra/prow/plugins/close"
```
Comment (apelisse) [inline prow/cmd/hook/main.go]: This is kind of unfortunate that we have to add that every single time :-/

File: prow/github/client.go [hunk 2]
```diff
@@ -253,3 +253,18 @@ func (c *Client) RemoveLabel(owner, repo string, number int, label string) error
 	}
 	return nil
 }
+
+func (c *Client) CloseIssue(owner, repo string, number int) error {
```
Comment (apelisse) [inline prow/github/client.go]: You made the assumption that you "have everything you need from github" as an argument against go-github. It didn't hold true for more than half a day :-/

File: prow/plugins/close/close.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+const pluginName = "close"
+
+var closeRe = regexp.MustCompile(`(?mi)^\/close\r?$`)
+
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+}
+
+type githubClient interface {
```
Comment (apelisse) [inline prow/plugins/close/close.go]: This is very confusing. There is githubClient, gitHubClient, GithubClient, githubClient (again?).

File: prow/cmd/hook/main.go (hunk 1 above)
Comment (spxtr) [inline prow/cmd/hook/main.go]: There's a unit test if you forget to add it and try to add it to `plugins.yaml`. The nice thing is that if you don't want to include a plugin in the binary you can just not include it here.

(I wanted to put these lines in the `plugins` package but it makes an import cycle :P)

Comment (spxtr) [inline prow/cmd/hook/main.go]: But I agree, it's a little annoying. Hopefully not too bad.

File: prow/plugins/close/close.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		return nil
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	number := ic.Issue.Number
+
+	// Allow assignees and authors to close issues.
+	canClose := false
+	for _, assignee := range ic.Issue.Assignees {
```
Comment (apelisse) [inline prow/plugins/close/close.go]: Don't we have `IsAssignee()` already?

File: prow/plugins/close/close.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		if ic.Comment.User.Login == assignee.Login {
+			canClose = true
+			break
+		}
+	}
+	if ic.Comment.User.Login == ic.Issue.User.Login {
+		canClose = true
+	}
+
+	if !canClose {
```
Comment (apelisse) [inline prow/plugins/close/close.go]: Would have a much better semantic with:
```
if !IsAssignee(...) && !IsAuthor(...) {
    return comment(...)
}
return close(...)
```

I think.

**Comment by grodrigues3** [comment]
cc @kubernetes/contributor-experience

File: prow/plugins/close/close.go (hunk 3 above)
Comment (spxtr) [inline prow/plugins/close/close.go]: Fixed.

File: prow/plugins/close/close.go (hunk 5 above)
Comment (spxtr) [inline prow/plugins/close/close.go]: Done. Looks much nicer now.

File: prow/plugins/close/close.go (hunk 4 above)
Comment (spxtr) [inline prow/plugins/close/close.go]: Yes, I'll move it to the appropriate place.

**Comment by spxtr** [comment]
@apelisse thanks for the comments, PTAL.

File: prow/Makefile [hunk 6]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-HOOK_VERSION   = 0.45
+HOOK_VERSION   = 0.46
```
Comment (apelisse) [inline prow/Makefile]: Nothing with this PR in particular, but that seems like a pretty bad idea (though I know someone suggested the same to me)

**Comment by spxtr** [comment]
https://github.com/kubernetes/test-infra/pull/997#event-850355382 Whaaa? We don't have an approval-to-lgtm munger or plugin do we? How did this happen?

**Comment by krzyzacy** [comment]
oh woops I noticed this a while back and thought it was the designated behavior..

**Comment by apelisse** [comment]
Yeah, we have much more complex ACLs than what github allows :-)

**Comment by spxtr** [comment]
https://github.com/kubernetes/test-infra/issues/601 :)

**Pull request merged**